### ISSUE TITLE
Expose internal injector cache for cleanup

### ISSUE THREAD

**Issue by jutaz** [angular/angular.js#14319]
**Do you want to request a *feature* or report a *bug*?**

Feature that allows solving memory leak issues using `angular.injector()`.

**What is the current behavior?**

Currently `angular.injector()` [stores cache internally](https://github.com/angular/angular.js/blob/90cb24e7818154b2293c8257f115cd9b2924a38c/src/auto/injector.js#L641-L665), and does not expose it to outside world, meaning that if there is a reference to one of the items, that are sitting in cache, this whole gargantuan cache object sits in memory.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

This problem is really well described by my colleague [here](https://atticuswhite.com/blog/angular-unit-test-memory-problems/).

TL;DR - the memory grows and grows with each `inject` call. After some tracing I found out that the problem was massive amounts of caches still being retained by `angular.injector()`.

Using our app's test suite (karma + mocha + phantomjs) (~1000 mixed service, directive and controller tests) and [`memusg`](https://gist.github.com/netj/526585) utility, I measured that test process (`PhantomJS`) climbs up to `~3.2Gb` and crashes since PhantomJS simply runs out of memory, while with my simple patch it takes just ~`2.2GB`.

**What is the expected behavior?**

Expected behavior would be that `angular.injector()` could either cleanup after itself, or better yet, expose a function (`$$cleanup()`, for example) that would allow user to clean up injector and it's cache 

**What is the motivation / use case for changing the behavior?**

To allow running big codebases under a single browser instance (currently we have to split whole codebase into 8 chunks that run sequentially) and to improve memory usage of tests in general. I do not think this will directly benefit apps themselves.

**Which versions of Angular, and which browser / OS are affected by this issue? Did this work in previous versions of Angular? Please also test with the latest stable and snapshot (https://code.angularjs.org/snapshot/) versions.**

We are running angular `1.4.6` and corresponding version of `angular-mocks`, but I believe that newer versions have this problem as well, as that code was really barely touched.

`1.9.8` for phantomJS.

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**

Fixing this is quite simple - Either expose caches as properties or provide cleanup function in [here](https://github.com/angular/angular.js/blob/90cb24e7818154b2293c8257f115cd9b2924a38c/src/auto/injector.js#L882-L890). Then frameworks like `angular-mocks` will be able to delete injector cache, and it will be seamless to the end users too. Here is a quick code solution for `angular-mocks` (after exposing caches as properties). Assume code is added somewhere in [here](https://github.com/angular/angular.js/blob/master/src/ngMock/angular-mocks.js#L2761-L2775):
```js
Object.keys(injector.$$providerCache).forEach(function (key) {
  delete injector.$$providerCache[key];
});
Object.keys(injector.$$instanceCache).forEach(function (key) {
  delete injector.$$instanceCache[key];
});
```

I'm happy to submit patches or provide any helpful information in regards to solving this issue :+1:

**Comment by gkalpak** [comment]
The question is: Who is holding references to injector cache contents (and why) ?
(OK, maybe not _the_ question, but it's _a_ question, no doubt :smiley:)

**Comment by jutaz** [comment]
> The question is: Who is holding references to injector cache contents (and why) ?

Ooh, I would love to know that, but it is either within angular itself, or within `angular-mocks`. I would lean towards angular itself, but I have no evidence on this, unfortunately.

**Comment by jutaz** [comment]
Any ideas for this yet? I think this is a problem worth solving in spite of big codebases that have quite a lot of tests

**Comment by lgalfaso** [comment]
Without a way to reproduce this, it is very hard to move forward. I try to reproduce this with Jasmine without success. I am wondering if this is a mocha specific issue.

The way the resources should be released works this way:
- Add a global `afterEach` https://github.com/angular/angular.js/blob/master/src/ngMock/angular-mocks.js#L2792
- This global `afterEach` calls a function that releases the resources https://github.com/angular/angular.js/blob/master/src/ngMock/angular-mocks.js#L2738
https://github.com/angular/angular.js/blob/master/src/ngMock/angular-mocks.js#L2744


If someone creates a plunker that shows this memory leak, then someone can take a look.

**Comment by jutaz** [comment]
Okay, I can create a plunker, sure

**Comment by jutaz** [comment]
Here is the plunker - http://plnkr.co/edit/gNuw2r2kEL1k8Bx5DbIP

However, I recommend running this in `karma` to better see the memory usage of the actual tests VS whole webpage

**Comment by lgalfaso** [comment]
It looks like this is a mocha problem, in fact, the issue is described in the same article here
https://atticuswhite.com/blog/angular-unit-test-memory-problems/#comment-2333781362

I do not have a good solution, but if someone wants to look into this and post a PR to work around mocha's bug, then I can take a look.

**Comment by gkalpak** [comment]
I'm going to close this, since this is an issue with mocha, not Angular.
